// anime-actions/lib/functions.js
const resolve = require('./resolve.js')

module.exports = {

    baka: function () {
        return resolve.get('baka');
    },
    blush: function () {
        return resolve.get('blush')
    },
    cuddle: function () {
        return resolve.get('cuddle')
    },
    bite: function () {
        return resolve.get('bite')
    },
    dance: function () {
        return resolve.get('dance')
    },
    slap: function() {
        return resolve.get('slap')
    },
    bonk: function() {
        return resolve.get('bonk')
    },
    bully: function() {
        return resolve.get('bully')
    },
    hug: function() {
        return resolve.get('hug')
    },
    kiss: function() {
        return resolve.get('kiss')
    },
    pat: function() {
        return resolve.get('pat')
    },
    wallpaper: function() {
        return resolve.get('wallpaper')
    },
    happy: function() {
        return resolve.get('happy')
    },
    wink: function() {
        return resolve.get('wink')
    },
    kill: function() {
        return resolve.get('kill')
    },
    highfive: function() {
        return resolve.get('highfive')
    },
    wave: function() {
        return resolve.get('wave')
    },
    cry: function() {
        return resolve.get('cry')
    },
    smile: function() {
        return resolve.get('smile')
    },
    punch: function() {
        return resolve.get('punch')
    },
    kick: function() {
        return resolve.get('kick')
    },
    poke: function() {
        return resolve.get('poke')
    },
    goodnight: function() {
        return resolve.get('goodnight')
    },
    zerotwo: function() {
        return resolve.get('zerotwo')
    },
    confused: function() {
        return resolve.get('confused')
    },
    sad: function() {
        return resolve.get('sad')
    },
    thinking: function() {
        return resolve.get('thinking')
    },
    yes: function() {
        return resolve.get('yeah')
    },
    yeet: function() {
        return resolve.get('yeet')
    },
    stare: function() {
        return resolve.get('stare')
    },
    bored: function() {
        return resolve.get('bored')
    },
    scream: function() {
        return resolve.get('scream')
    },
    nervous: function() {
        return resolve.get('nervous')
    }
}
